File: csharp/HomeScheduler/HomeScheduler.Service/Service/UserService.cs
using HomeScheduler.Common.Dto;
using HomeScheduler.Data;
using HomeScheduler.Data.Entity;
using HomeScheduler.Data.Repository;

namespace HomeScheduler.Service.Service;

public class UserService : IUserService
{
    private readonly IUserRepository _repo;

    public UserService(IUserRepository repo)
    {
        _repo = repo;
    }

    public List<UserDto> GetUsers()
    {
        List<UserDto> result = _repo.GetUsers().Select(Map).ToList();
        return result;
    }

    public UserDto? GetUser(int id)
    {
        UserDto? result = Map(_repo.GetUser(id));
        return result;
    }

    public UpdateResponseDto<UserDto> AddUser(RequestDto<UserRequestDto> dto)
    {
        User entity = Map(dto);
        UpdateResponse<User> result = _repo.AddUser(entity);
        return new UpdateResponseDto<UserDto>
        {
            Data = Map(result.Data),
            ChangeCount = result.ChangeCount
        };
    }

    public UpdateResponseDto<UserDto> UpdateUser(int id, RequestDto<UserRequestDto> dto)
    {
        User entity = Map(dto);
        UpdateResponse<User> result = _repo.UpdateUser(id, entity);
        return new UpdateResponseDto<UserDto>
        {
            Data = Map(result.Data),
            ChangeCount = result.ChangeCount
        };
    }

    private User Map(RequestDto<UserRequestDto> dto)
    {
        DateTime dt = DateTime.UtcNow;

        return new User
        {
            UserName = dto.Data.UserName,
            FirstName = dto.Data.FirstName,
            LastName = dto.Data.LastName,
            CreatedBy = dto.UpdatedBy,
            CreatedDate = dt,
            UpdatedBy = dto.UpdatedBy,
            UpdatedDate = dt
        };
    }

    private UserDto? Map(User? entity)
    {
        if (entity == null) return null;

        return new UserDto
        {
            UserId = entity.UserId,
            UserName = entity.UserName,
            FirstName = entity.FirstName,
            LastName = entity.LastName,
            CreatedBy = entity.CreatedBy,
            CreatedDate = entity.CreatedDate,
            UpdatedBy = entity.UpdatedBy,
            UpdatedDate = entity.UpdatedDate
        };
    }
}

File: python-django/HomeScheduler/api/service/UserService.py
import json
from datetime import datetime

from ..models import User
from ..data.UserRepository import UserRepository
from ..dto.RequestDto import RequestDto
from ..dto.UserDto import UserDto
from ..dto.UserRequestDto import UserRequestDto
from ..dto.UpdateResponseDto import UpdateResponseDto


class UserService:
    repo: UserRepository = UserRepository()

    def get_users(self):
        # https://stackabuse.com/map-filter-and-reduce-in-python-with-examples/
        result = list(map(lambda x: self.map_to_dto(x), self.repo.get_users()))
        return result

    def get_user(self, user_id):
        result = self.map_to_dto(self.repo.get_user(user_id))
        return result

    def add_user(self, dto: RequestDto[UserRequestDto]) -> UpdateResponseDto[UserDto]:
        entity = self.map_to_entity(dto)
        result = self.repo.add_user(entity)
        return UpdateResponseDto(self.map_to_dto(result.data), result.changeCount)

    def update_user(self, user_id, dto: RequestDto[UserRequestDto]):
        entity = self.map_to_entity(dto)
        result = self.repo.update_user(user_id, entity)
        return UpdateResponseDto(self.map_to_dto(result.data), result.changeCount)

    def map_to_dto(self, entity: User):
        return UserDto(
            userId=entity.userid,
            userName=entity.username,
            firstName=entity.firstname,
            lastName=entity.lastname,
            createdBy=entity.createdby,
            createdDate=entity.createddate,
            updatedBy=entity.updatedby,
            updatedDate=entity.updateddate)

    def map_to_entity(self, dto: RequestDto[UserRequestDto]):
        dt = datetime.utcnow()

        entity = User(
            username=dto.data.userName,
            firstname=dto.data.firstName,
            lastname=dto.data.lastName,
            createdby=dto.updatedBy,
            createddate=dt,
            updatedby=dto.updatedBy,
            updateddate=dt)
        return entity
